### PR TITLE
Adds Servo Power Instrument

### PR DESCRIPTION
**Pull request by vflouris**

### CODE REVIEW

File: wlauto/instrumentation/servo_power_monitors/__init__.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+from wlauto.exceptions import InstrumentError, ConfigError
+from wlauto.utils.types import list_of_strings
+from wlauto.utils.misc import check_output
+from wlauto.utils.cros_sdk import CrosSdkSession
+
+
+class ServoPowerMonitor(Instrument):
+
+    name = 'servo_power_monitor'
+    description = """Collects power traces using the Servo Board.
```
Comment (setrofim) [inline wlauto/instrumentation/servo_power_monitors/__init__.py]: s/Servo Board/Chromium OS Servo Board/

File: wlauto/instrumentation/servo_power_monitors/__init__.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+        self.poller.start()
+        self.stopped = False
+
+    def stop(self, context):
+        self.power_data = self.poller.stop()
+        self.poller.join()
+        self.stopped = True
+
+    def update_result(self, context):
+        os.mkdir(os.path.dirname(self.outfiles[0]))
```
Comment (setrofim) [inline wlauto/instrumentation/servo_power_monitors/__init__.py]: see `wlauto.utils.misc.ensure_file_dictory_exists`

File: wlauto/instrumentation/servo_power_monitors/__init__.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+# pylint: disable=W0613,E1101,attribute-defined-outside-init
```
Comment (setrofim) [inline wlauto/instrumentation/servo_power_monitors/__init__.py]: `from __future__ import division` 

(there is a bunch of devision in the result processing which assumes floats; this makes it safer)

File: wlauto/instrumentation/servo_power_monitors/__init__.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        for i, outfile in enumerate(self.outfiles):
+            self.power_data[i] = [float(v) / 1000.0 for v in self.power_data[i]]
+            metric_name = '{}_power'.format(self.labels[i])
+            sample_sum = sum(self.power_data[i])
+            power = sample_sum / len(self.power_data[i])
+            context.result.add_metric(metric_name, round(power, 3), 'Watts')
+            metric_name = '{}_energy'.format(self.labels[i])
+            energy = sample_sum * (1.0 / self.sampling_rate)
+            context.result.add_metric(metric_name, round(energy, 3), 'Joules')
+            with open(outfile, 'w+') as f:
```
Comment (setrofim) [inline wlauto/instrumentation/servo_power_monitors/__init__.py]: Why `'w+'` mode here? I would expect `'wb'`.

File: wlauto/instrumentation/servo_power_monitors/__init__.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+            self.cros_session.send_command(self.command)
+            lines = self.cros_session.get_lines()
+            for i, line in enumerate(lines):
+                _, value = line.split(':')
+                self.power_data[i].append(value)
+        self.term_signal.set()
+
+    def stop(self):
+        self.term_signal.clear()
+        self.term_signal.wait()
```
Comment (setrofim) [inline wlauto/instrumentation/servo_power_monitors/__init__.py]: This should be `self.wait()`

File: wlauto/instrumentation/servo_power_monitors/__init__.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+    def run(self):
+        while self.term_signal.is_set():
+            time.sleep(self.period)
+            self.cros_session.send_command(self.command)
+            lines = self.cros_session.get_lines()
+            for i, line in enumerate(lines):
+                _, value = line.split(':')
+                self.power_data[i].append(value)
+        self.term_signal.set()
```
Comment (setrofim) [inline wlauto/instrumentation/servo_power_monitors/__init__.py]: Why?

File: wlauto/instrumentation/servo_power_monitors/__init__.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+                  description="""Path to chroot direcory on the host."""),
+        Parameter('sampling_rate', kind=int, default=10,
+                  description="""Samples per second."""),
+        Parameter('board_name', kind=str, mandatory=True,
+                  description="""The name of the board under test."""),
+    ]
+
+    def initialize(self, context):
+        # pylint: disable=access-member-before-definition
+        self.in_chroot = False if subprocess.call('which dut-control', stdout=subprocess.PIPE, shell=True) else True
```
Comment (setrofim) [inline wlauto/instrumentation/servo_power_monitors/__init__.py]: See `wlauto.utils.misc.which`

File: wlauto/utils/cros_sdk.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+            self.queue.put(line)
+
+    def set_stop(self):
+        self.stop_signal = True
+
+
+class CrosSdkSession(object):
+
+    def __init__(self, cros_path):
+        self.in_chroot = False if subprocess.call('which dut-control', stdout=subprocess.PIPE, shell=True) else True
```
Comment (setrofim) [inline wlauto/utils/cros_sdk.py]: See `wlauto.utils.misc.which`

File: wlauto/utils/cros_sdk.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+        ON_POSIX = 'posix' in sys.builtin_module_names
+        if self.in_chroot:
+            self.cros_sdk_session = Popen(['/bin/bash'], bufsize=1, stdin=PIPE, stdout=PIPE, stderr=PIPE,
```
Comment (setrofim) [inline wlauto/utils/cros_sdk.py]: `'/bin/sh'` ?

File: wlauto/utils/cros_sdk.py [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+        self.stderr_queue = Queue()
+        self.stderr_thread = OutputPollingThread(self.cros_sdk_session.stderr, self.stderr_queue, 'stderr')
+        self.stderr_thread.daemon = True
+        self.stderr_thread.start()
+
+    def kill_session(self):
+        self.stdout_thread.set_stop()
+        self.stderr_thread.set_stop()
+        self.send_command('echo foo >&1')  # send something into stdout to unblock it and close it properly
+        self.send_command('echo foo 1>&2')  # ditto for stderr
```
Comment (setrofim) [inline wlauto/utils/cros_sdk.py]: `echo TERM_SIGNAL` or something less arbitrary than 'foo'.

File: wlauto/utils/cros_sdk.py [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+        self.cros_sdk_session.kill()
+
+    def send_command(self, cmd, flush=True):
+        if not cmd.endswith('\n'):
+            cmd = cmd + '\n'
+        self.cros_sdk_session.stdin.write(cmd)
+        if flush:
+            self.cros_sdk_session.stdin.flush()
+
+    def read_line(self, timeout_in_ms=0):
```
Comment (setrofim) [inline wlauto/utils/cros_sdk.py]: Why specify timeout in milliseconds only convert to seconds at the place it is actually used? (also, every other timeout in WA is in seconds).

File: wlauto/utils/cros_sdk.py [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+        return _read_line_from_queue(self.stdout_queue, timeout_in_ms=timeout_in_ms)
+
+    def read_stderr_line(self, timeout_in_ms=0):
+        return _read_line_from_queue(self.stderr_queue, timeout_in_ms=timeout_in_ms)
+
+    def get_lines(self, timeout_in_ms=0, timeout_only_for_first_line=True, from_stderr=False):
```
Comment (setrofim) [inline wlauto/utils/cros_sdk.py]: Have you considered using fdpexect instead of this?

**Comment by freedomtan** [comment]
Please do use chroot() to cros sdk, it takes time (more than one second on my machine). I think my second trial which uses xmlrpc is better. https://github.com/freedomtan/workload-automation/commit/de0b7bd1487b61a108029b382a2bcccde1e1bd8f

File: wlauto/instrumentation/servo_power_monitors/__init__.py (hunk 5 above)
Comment (setrofim) [inline wlauto/instrumentation/servo_power_monitors/__init__.py]: Sorry, that's `self.join()`, not `wait()`

**Comment by setrofim** [comment]
Obsoleted by https://github.com/ARM-software/workload-automation/pull/144

**Pull request closed without merging**